### ISSUE TITLE
Minio unresponsive after shutting down node in a 4 node docker swarm

### ISSUE THREAD

**Issue by roelio** [minio/minio#5092]
Created a 4 node cluster like https://github.com/minio/minio/tree/master/docs/orchestration/docker-swarm with a minio container pinned on each node. A traefik loadbalancer docker service in put in front of them.

When shutdown a node the minio system becomes unresponsive. When I just kill a minio container (not the whole host) everything works as expected. I've tested this through the load balancer and circumventing the load balancer by directly connecting with one of the alive nodes. I used the version from the guide and also the edge version, both did not work

It looks similar to the issue https://github.com/minio/minio/issues/4887 (and maybe also related to https://github.com/minio/minio/issues/5089), but the circuit-breaker workaround didnt work for me.

Slack messages: https://minio.slack.com/archives/C3NDUB8UA/p1508502645000199

**Comment by vadmeste** [comment]
Minio log as seen by @roelio:

`time="2017-10-20T12:46:32Z" level=error msg="Unable to fetch disk info for &cmd.retryStorage{remoteStorage:(*cmd.networkStorage)(0xc42000c080), maxRetryAttempts:1, retryUnit:1000000, retryCap:5000000, offline:true, offlineTimestamp:time.Time{sec:63644100345, nsec:549945278, loc:(*time.Location)(nil)}}" cause="disk not found" source="[xl-v1.go:199:getDisksInfo()]"`


`time="2017-10-20T12:45:45Z" level=error msg="Unable to get server info from minio3:9000." cause="dial-http minio3:9000/minio/admin: dial tcp: lookup minio3 on 127.0.0.11:53: no such host" source="[admin-handlers.go:284:adminAPIHandlers.ServerInfoHandler.func1()]" ` (edited)

**Comment by vadmeste** [comment]
I am going to test this and report.

**Comment by roelio** [comment]
On the edge version I dont get the above log messages.

**Comment by vadmeste** [comment]
@roelio those logs are normal (they are like verbose log), they are shown because you called `mc admin info command` when one of the nodes is down.

I am testing directly with Swarm right now. When I shutdown one VM, Minio service kept restarting itself but I am not sure why. I will investigate more.

**Comment by harshavardhana** [comment]
Turn off health check @vadmeste are you not using edge ?

**Comment by roelio** [comment]
@vadmeste I also had disabled the health checks because of the restarting issue

**Comment by roelio** [comment]
I can still reach the website, but cannot download, upload or do anything and get a 504 form ajax request that should normally retrieve the available size etc.

![image](https://user-images.githubusercontent.com/900675/31835927-d575e6e2-b5d3-11e7-95a7-5ed08d347952.png)

**Comment by harshavardhana** [comment]
Listing is allowed by looking up the disks which are available @roelio so due to random chance you are able to List as well.  For downloads and uploads the disk is supposed to return an appropriate error for it to be ignored, since there are no deadlines on RPC connection we will have sort of this type of problem of the underlying system didn't respond correctly or just hanging. 

@vadmeste migrating to GRPC will help here and we can avoid underlying networking issues by having proper deadlines.

**Comment by davidgeorgealex** [comment]
I'm encountering the same issue, but in an OpenShift environment using AWS instances. When I kill just the container (Docker), the cluster is accessible, but when I bring down the whole AWS instance, I'm facing unresponsiveness issue.

**Comment by harshavardhana** [comment]
How was the system taken down @davidgeorgealex ? it is a clean shutdown system shutdown like using `reboot` or `shutdown -f` ?  or not?

**Comment by davidgeorgealex** [comment]
@harshavardhana We have used AWS UI to stop the instance. I would say it was a clean shutdown.

**Comment by vadmeste** [comment]
I was able to reproduce this with Vagrant + Swarm.

When a Vagrant VM is shutdown, Swarm creates a new docker container as a replacement to the failed one in another VM, to keep having 4 total instances.

However, the new Minio instance doesn't fully start:

```
time="2017-10-25T16:42:34Z" level=error msg="Unable to resolve host minio3 (21 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
time="2017-10-25T16:42:35Z" level=error msg="Unable to resolve host minio3 (22 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
time="2017-10-25T16:42:41Z" level=error msg="Unable to resolve host minio3 (28 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
Created minio configuration file successfully at /root/.minio
New disk(s) were found, format them by running - "mc admin heal myminio"

[01/04] http://minio1:9000/export - 9.6 GiB online
[02/04] http://minio2:9000/export - 9.6 GiB online
[03/04] http://minio3:9000/export - 9.6 GiB online
[04/04] http://minio4:9000/export - 9.6 GiB online
time="2017-10-25T17:13:11Z" level=error msg="Failed to delete lock info entry (param = {testbucket }; opsID = JLATJ26U0C0P7J0Z)" cause="No entry in lock info for <Operation ID> JLATJ26U0C0P7J0Z, <volume> testbucket, <path> " source="[namespace-lock.go:184:(*nsLockMap).lock()]" stack="lock-instrument.go:249:(*nsLockMap).deleteLockInfoEntryForOps namespace-lock.go:182:(*nsLockMap).lock namespace-lock.go:331:(*lockInstance).GetLock xl-v1-healing.go:94:healBucket xl-v1-healing.go:321:quickHeal xl-v1.go:162:newXLObjects xl-v1.go:73:newXLObjectLayer server-main.go:260:newObjectLayer server-main.go:207:serverMain /go/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:499:github.com/minio/minio/vendor/github.com/minio/cli.HandleAction /go/src/github.com/minio/minio/vendor/github.com/minio/cli/command.go:214:github.com/minio/minio/vendor/github.com/minio/cli.Command.Run /go/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:260:github.com/minio/minio/vendor/github.com/minio/cli.(*App).Run main.go:145:Main /go/src/github.com/minio/minio/main.go:68:main.main /usr/local/go/src/runtime/proc.go:185:runtime.main /usr/local/go/src/runtime/asm_amd64.s:2337:runtime.goexit" 
time="2017-10-25T17:13:11Z" level=fatal msg="Unable to initialize XL object layer." cause="Operation timed out: " source="[xl-v1.go:74:newXLObjectLayer()]" 

```

This is an edge docker image
```
/go/src/github.com/minio # minio version
Version: 2017-10-25T02:06:11Z
Release-Tag: DEVELOPMENT.2017-10-25T02-06-11Z
Commit-ID: e94d64568fd7ec27e3e94706e925f229c60bc046
```

Actually, I failed to reproduce this issue manually: creating local instances in my machine, killing one instance, removing data related to that instance and starting the instance again.

I guess it is something with the distributed lock.. in combination with another problem.

**Comment by harshavardhana** [comment]
The problem is host resolution should wait. It is not waiting and failing. New swarm node will take time to come online.

**Comment by vadmeste** [comment]
@harshavardhana, in practice the new swarm node is quickly created in another VM (let's say after few seconds).. but the log of the new docker instance only shows this:

```
time="2017-10-25T16:42:34Z" level=error msg="Unable to resolve host minio3 (21 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
time="2017-10-25T16:42:35Z" level=error msg="Unable to resolve host minio3 (22 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
time="2017-10-25T16:42:41Z" level=error msg="Unable to resolve host minio3 (28 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
Created minio configuration file successfully at /root/.minio
New disk(s) were found, format them by running - "mc admin heal myminio"

[01/04] http://minio1:9000/export - 9.6 GiB online
[02/04] http://minio2:9000/export - 9.6 GiB online
[03/04] http://minio3:9000/export - 9.6 GiB online
[04/04] http://minio4:9000/export - 9.6 GiB online
```

and gets stuck there for many minutes

**Comment by harshavardhana** [comment]
> @harshavardhana, in practice the new swarm node is quickly created in another VM (let's say after few seconds).. but the log of the new docker instance only shows this:
> 
> time="2017-10-25T16:42:34Z" level=error msg="Unable to resolve host minio3 (21 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
> time="2017-10-25T16:42:35Z" level=error msg="Unable to resolve host minio3 (22 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
> time="2017-10-25T16:42:41Z" level=error msg="Unable to resolve host minio3 (28 seconds elapsed)" cause="lookup minio3 on 127.0.0.11:53: no such host" source="[net.go:101:getHostIP4()]" 
> Created minio configuration file successfully at /root/.minio
> New disk(s) were found, format them by running - "mc admin heal myminio"
> 
> [01/04] http://minio1:9000/export - 9.6 GiB online
> [02/04] http://minio2:9000/export - 9.6 GiB online
> [03/04] http://minio3:9000/export - 9.6 GiB online
> [04/04] http://minio4:9000/export - 9.6 GiB online
> and gets stuck there for many minutes

Oh this looks like some other problem but it is not this issue IMO @vadmeste can you send a `SIGABRT` and crash the server. We can see where its waiting.

**Comment by vadmeste** [comment]
http://paste.ubuntu.com/25818694/

I kept the cluster running for ~ 20 minutes, then I invoked ls command, so we should focus on go routines with ~ 2 minutes wait time I guess

**Comment by harshavardhana** [comment]
> http://paste.ubuntu.com/25818694/
> 
> I kept the cluster running for ~ 20 minutes, then I invoked ls command, so we should focus on go routines with ~ 2 minutes wait time I guess

From what it looks like 21minutes wait is happening inside http2 frame reader, can you test disabling HTTP2 here? Like starting server with. 

```
GODEBUG=h2client=0
GODEBUG=h2server=0
```

**Comment by roelio** [comment]
@vadmeste if you put placement constraints you can make sure the container does not start on another node. I labeled my nodes and set the constraints as follows:
```
version: '3.2'

# A Minio cluster of with 4 instances
# Minio is available on the services network with hostname minio ergo minio.services
# Two proxies on the swarm will redirect the data to the minio instances

# The deploy command including: docker stack deploy -c minio.yml minio
# Environment variables required: NODE1_IDENTIFIER, NODE2_IDENTIFIER, NODE3_IDENTIFIER, NODE4_IDENTIFIER
services:
  minio1:
    image: minio/minio:edge
    healthcheck:
      disable: true
    volumes:
      - minio1-data:/export
    deploy:
      placement:
        constraints: [node.labels.name == alfa]
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 120s
    networks:
      minio:
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    secrets:
      - secret_key
      - access_key

  minio2:
    image: minio/minio:edge
    healthcheck:
      disable: true
    volumes:
      - minio2-data:/export
    deploy:
      placement:
        constraints: [node.labels.name == beta]
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 120s
    networks:
      minio:
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    secrets:
      - secret_key
      - access_key

  minio3:
    image: minio/minio:edge
    healthcheck:
      disable: true
    volumes:
      - minio3-data:/export
    deploy:
      placement:
        constraints: [node.labels.name == gamma]
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 120s
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    networks:
      minio:
    secrets:
      - secret_key
      - access_key

  minio4:
    image: minio/minio:edge
    healthcheck:
      disable: true
    volumes:
      - minio4-data:/export
    deploy:
      placement:
        constraints: [node.labels.name == delta]
      restart_policy:
        delay: 10s
        max_attempts: 10
        window: 120s
    command: server http://minio1/export http://minio2/export http://minio3/export http://minio4/export
    networks:
      minio:
    secrets:
      - secret_key
      - access_key

volumes:
  minio1-data:

  minio2-data:

  minio3-data:

  minio4-data:

secrets:
  secret_key:
    external: true
  access_key:
    external: true

networks:
  minio:
    driver: overlay 
    ipam:
      driver: default
      config:
        - subnet: 10.0.111.0/26

```

**Comment by davidgeorgealex** [comment]
Hi guys, do you have an update on this? I'm still facing issues with a Minio cluster deployed on AWS (3 EC2, each with 1 Minio instance). If I stop an instance (from AWS console), then the cluster becomes unresponsive (tried to run `mc ls <<minio-instance>>). I get a timeout, but after around 20mins of waiting. Also, after around 40mins it seams like Minio see that the instance is offline and the cluster is working again and shows that 1 instance is offline. If I just start back the EC2, then everything is working as expected. I'm wondering if it's a problem with `docker events` mechanism. If there any action that I can perform to gather more details about the issue?

**Comment by davidgeorgealex** [comment]
Another interesting fact is that I'm not able to run `mc ls play`, even if `play` is the default env. Also if I run `mc config host ls` I get `play :  https://play.minio.io:9000      Q3AM3UQ867SPQQA43P2F  zuf+tfteSlswRu7BJ86wekitnifILbZam1KYY3TG  S3v4`.

Please note that I'm using Docker, so my EC2 instances contains a docker container.
Minio version: RELEASE.2017-11-22T19-55-46Z

**Comment by davidgeorgealex** [comment]
Tried to stop the AWS instance using `shutdown` as well and it's the same behavior (unresponsive cluster). If I just kill the docker container that sits on that instance (`docker kill`), the cluster continue to work

**Comment by vadmeste** [comment]
@davidgeorgealex, can you open a new issue? it is probably easier to examine your use case first. Can you paste your docker-compose file ?

**Comment by vadmeste** [comment]
Well, I am able to reproduce this with vagrant and without docker or docker swarm.

`mc ls` doesn't respond, although a simple curl command is responsive.

`mc admin lock --duration 0s myminio/testbucket` also shows some waiting locks.

```
15:00 $ mc admin lock list --duration 0s myminio/testbuket/
testbuket/ (LocksOnObject: 0, locksAcquiredOnObject: 0, totalBlockLocks:0):   {OperationID:7b5738ca-8023-4b53-93ad-15184a0450ce LockSource:[bucket-handlers.go:658:objectAPIHandlers.HeadBucketHandler()] LockType:RLock Status:Blocked Since:2017-12-15 13:59:04.059323327 +0000 UTC}

testbuket/ (LocksOnObject: 0, locksAcquiredOnObject: 0, totalBlockLocks:0):   {OperationID:7b5738ca-8023-4b53-93ad-15184a0450ce LockSource:[bucket-handlers.go:658:objectAPIHandlers.HeadBucketHandler()] LockType:RLock Status:Blocked Since:2017-12-15 13:59:04.059323327 +0000 UTC}  {OperationID:3a259071-88bc-46c1-90ec-a247477862ff LockSource:[bucket-handlers.go:658:objectAPIHandlers.HeadBucketHandler()] LockType:RLock Status:Blocked Since:2017-12-15 13:59:22.105205135 +0000 UTC}
```

This is probably a dsync issue.. calling the failing node probably doesn't return an error only after a long timeout. Will check more and update.

**Comment by vadmeste** [comment]
I am using the latest Minio release. DSYNC_LOG is enabled.

The cluster gets stuck.. for long minutes.. when it gets back.. the following logs are printed..

```
2017/12/15 15:04:39 Unable to call RLock read tcp 192.168.14.10:50114->192.168.14.13:9000: read: no route to host
2017/12/15 15:04:39 Unable to call RLock read tcp 192.168.14.10:50114->192.168.14.13:9000: read: no route to host
2017/12/15 15:04:39 Unable to call RLock read tcp 192.168.14.10:50114->192.168.14.13:9000: read: no route to host
```
Then the normal cluster behavior is back, 

Now it prints similar error messages, which is a normal thing, since the failed node is still offline.

```
2017/12/15 15:04:44 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
2017/12/15 15:04:47 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
2017/12/15 15:04:50 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
2017/12/15 15:05:02 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
2017/12/15 15:05:05 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
2017/12/15 15:05:30 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
2017/12/15 15:05:33 Unable to call RLock dial-http 192.168.14.13:9000/minio/lock/tmp/m4: dial tcp 192.168.14.13:9000: getsockopt: no route to host
```

I never examined the code of dsync code before.. but I guess there is a problem there..

**Comment by vadmeste** [comment]
Also I guess `no route to host` is causing this.. because we probably only tested with `Connection refused` error or probably other network errors.

**Comment by vadmeste** [comment]
I am much closer now.

If I comment the following `if` block: https://github.com/minio/minio/blob/8c08571cd9e6c69cc911f88828b174efdd9caef1/cmd/net-rpc-client.go#L62

then the problem will disappear.

Need more investigation

**Comment by vadmeste** [comment]
I really thought this change can fix the problem.. but it didn't..

https://github.com/vadmeste/minio/commit/29d799b2be0bce90e96499ba399b1ac46f410d58

Will update tomorrow on this.

**Comment by harshavardhana** [comment]
@vadmeste is there anything that you found here ?

**Comment by harshavardhana** [comment]
> @vadmeste is there anything that you found here ?

ping @vadmeste

**Comment by harshavardhana** [comment]
The PR is merged and this issue should be fixed as of #5840

**Comment by roelio** [comment]
Awesome work guys :100: 

Will give it another try soon :+1:
